R/util_data_summaries.R

```diff
@@ -126,3 +155,509 @@ ds.summarize_by_column <- function(
         return()
 }
 
+#########################################################
+###
+###     ds.glm1 Functions
+###
+###     All specific to models that meet ds.glm1 
+###     model specification. See ds.validate_glm1_model()
+###
+#########################################################
+
+
+ds.validate_glm1_model <- function(glm_model){
+    # returns TRUE if model meets ds.glm1 spec, else FALSE
+    
+    #   canonical ds.glm1 formulas follow this pattern:
+    #       dv ~ iv [ * mod ] [ + cov1 + cov2 + ... ]
+    
+    # Why have a ds.glm1 spec?
+    #   The spec increases predictability and restricts
+    #   possible specifications to reduce summary complexity.
+    #   Summaries of multi-level variables and multi-way interactions
+    #   would be (will be) more challenging to code for.
+    
+    # ds.glm1 spec:
+    #   iv is first predictor
+    #   if moderator present, it is second predictor 
+    #       and "*" separates iv and moderator 
+    #   covs start at first "+" after iv
+    #   only 1 contrast per iv and moderator variable, specifically
+    #       iv and moderator vars must be boolean or numeric
+    #   up to 1 2-way interaction between iv and moderator
+    #   no other interaction terms can be specified via "*"
+    #       to use interactions in covariates, create manually
+    
+    varlist <- ds.helper$glm1_formula_to_varlist(glm_model$formula)
+    iv <- glm_model$model[,varlist$iv]
+    iv_var_type <- ds.helper$variable_type(iv)
+    is_valid <- FALSE
+    
+    # check iv suitability
+    if( iv_var_type %in% c("boolean","numeric") ){
+        is_valid <- TRUE
+    } else{
+        "ds.glm1 spec only allows ivs and moderators of type " %+% 
+            "boolean or numeric. Supplied iv is: "  %+% iv_var_type %+%
+            " in formula: " %+% glm_model$formula %>%
+            util.warn()
+    }
+    if( ! is.na(varlist$mod) ){ 
+        mod <- glm_model$model[,varlist$mod]
+        mod_var_type <- ds.helper$variable_type(mod)
+        if( ! mod_var_type %in% c("boolean","numeric") ){
+            "ds.extract_glm1_stats only supports mods of type " %+% 
+                "boolean or numeric. Supplied mod is: " %+% mod_var_type %+%
+                " in formula: " %+% glm_model$formula %>%
+                util.warn()
+            is_valid <- FALSE
+        }
+    }
+    return(is_valid)
+}
+
+
+ds.build_glm1_formulas <- function(
+    meta_formula = NULL, # @todo: enable specification through formula 
+    variable_vectors = NULL, # @todo: meta_formula can reference named vectors
```

Comment:
For now, I'm going to leave them first because I fantasize about ds.glm to eventually looking and feeling like glm (where the formula argument comes first). E.g., I think it'd be nice to call this like so, and have it parse the [] into arguments.

```r

ds.build_glm1_formulas([DV1, DV2] ~ [IV1, IV2] * MOD + [COV1, COV1 + COV2])


```